### ISSUE TITLE
Improvements to i3bar

### ISSUE THREAD

**Issue by infoseek12** [i3/i3#4314]
## I'm submitting a…

<pre>
[ ] Bug
[x] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Desired Behavior
I3bar is great but I think a few tweaks that could substantially improve it.

1. Ordering of the tray icons
At present, my understanding is that tray icons are always sorted alphabetically by name (I've seen some comments that they are sorted by the order they're initialized, but the reorder_trayclients_cmp in i3bar/src/xcb.c seems to put them in alphabetical order). It's a minor annoyance but it means that permanent applets (like Bluetooth and sound control) are sometimes displayed to the left of transient applications like VLC - contrary to the UI design of many systems trays and good semantic ordering. If there was a configuration option for tray ordering that allowed 4 options (alphabetical, reverse alphabetical, order initialized, and reverse order initialized) the ordering sought after in many use cases would be possible.

2. Smarter autohide behavior
It would be great if you could interact with applets in the system tray without holding down the mod key. I think the autohide mode would work better if it didn't immediately hide the bar upon release of the mod key if your cursor's either within the area of the menu bar or one the applets in the system tray has a menu open (at present, if you release the mod key the menu just floats there devoid of context). A "smart autohide" configuration option would be great and I think it might even be a better default behavior.

3. Ability to toggle autohide
Usually I don't want to waste any space displaying a systems bar when I'm not using it but there are situations when being able to see the i3bar, without holding down the mod key, would be really convenient. The ability to toggle between a fixed menu bar and the autohide mode, without changing the config file, would be really useful.

<pre>
[x] This feature requires new configuration and/or commands
</pre>

**Comment by i3bot** [comment]
Please note that new features which require additional configuration will usually not be considered. We are happy with the feature set of i3 and want to focus in fixing bugs instead. We do accept feature requests, however, and will evaluate whether the added benefit (clearly) outweighs the complexity it adds to i3.

**Comment by Airblader** [comment]
> reverse alphabetical

That's no more or less semantic, and doesn't address the issue you stated at all, does it?

> order initialized, and reverse order initialized

This used to be the behavior, and we changed it to alphabetical to make the order _predictable_. Ordering it based on which one "comes first" made for a significantly worse user experience given the race-condition-y nature of it.

> I think the autohide mode would work better if it didn't immediately hide the bar upon release of the mod key if your cursor's either within the area of the menu bar

I think this would upset heavy mouse users who don't care where their cursor is currently at.

> or one the applets in the system tray has a menu open

This sounds like a good idea, but unfortunately I don't think we have any (reasonable) way of knowing whether this is the case.

> A "smart autohide" configuration option would be great

That's gonna be a hard "no" for me, sorry, see i3bot's comment above.

> The ability to toggle between a fixed menu bar and the autohide mode, without changing the config file, would be really useful.

This feels a bit too edge-case-y for my taste, and I think editing the config file (automatically, if need be) is a good enough solution to this for me. Note that you can reload the config in-place and don't need to restart i3 altogether.

**Comment by infoseek12** [comment]
> reverse alphabetical
>> That's no more or less semantic, and doesn't address the issue you stated at all, does it?

No it would't. Initialization order would be the most useful option in my use case, as it would allow me to place the icons in an order that makes semantic sense. I only suggest the alphabetical options because of the present defaults.

>I think this would upset heavy mouse users who don't care where they cursor is currently at.

I'm not sure I understand this objection, do you mean keyboard users? In any case, even for heavy keyboard users interacting with applets in the system tray is necessarily preformed using a mouse as there aren't usually keyboard shortcuts. I can see how it might irritate users by not closing - I'm a pretty heavy keyboard user myself and like a lot of keyboard users, I think, I use Unclutter to hide the cursor when it's not it use. When Unclutter hides the inactive cursor in i3wm it doesn't draw focus back to the window the hidden cursor is above (most of the time). If this behavior was also true of mouse detection, when the cursor is within i3bar, then I think it might not be much of an issue for many users. Making this behavior an option would resolve the problem for anyone else that might have an issue.

>This sounds like a good idea, but unfortunately I don't think we have any (reasonable) way of knowing whether this is the case.

That's a shame. I'm afraid don't have enough knowledge to offer any suggestion. I understand why determining the behavior of third party components might be a challenge.

>This feels a bit too edge-case-y for my taste, and I think editing the config file (automatically, if need be) is a good enough solution to this for me. Note that you can reload the config in-place and don't need to restart i3 altogether.

True, it would just be a nice to have extra. And there is a possible solution so I can see why it's not a priority.

None of these are life and death issues but I do think they present opportunities for significant quality of life improvements. I3wm is great, by default or with a little tweaking, it does almost everything exactly how I want it to! These are just the walls I've personally run into, and see no way of scaling, in trying to get i3wm to behave the way I want it to.

**Comment by Airblader** [comment]
> Initialization order would be the most useful option in my use case

Just to clarify, this would mean starting them one by one _waiting_ for them to be mapped before starting the next one. Not just running the processes that creates them in a specific order.

And the repeating that everytime you restart i3, change monitor configuration etc. 

It really wasn't sane behavior, and alphabetical is the compromise we went with because we don't want configuration for this.

> I'm not sure I understand this objection, do you mean keyboard users? 

Yes, sorry, I misspoke.

> even for heavy keyboard users interacting with applets in the system tray is necessarily preformed using a mouse as there aren't usually keyboard shortcuts

That's true for the tray, but not the entire bar. Users might just have the mouse off to the top of the screen because they don't use it. If they show i3bar and then it stays open because of the cursor being there that would be an annoyance.

**Comment by infoseek12** [comment]
>Just to clarify, this would mean starting them one by one waiting for them to be mapped before starting the next one. Not just running the processes that creates them in a specific order.

>And the repeating that everytime you restart i3, change monitor configuration etc.

>It really wasn't sane behavior, and alphabetical is the compromise we went with because we don't want configuration for this.

Since all the applets for persistent services would be launched well in advance from the i3wm config file, at least in my use case, I don't think this would present a problem. By only allowing alphabetical ordering, as far as I can tell, there isn't any way to get the icons to line up in a correct semantic order or to have any kind of flexibility. I've seen other users asking about how to reorder system tray applets, by using initialization order (or providing that option) users would be able to have some control without developing a more complex solution that could explicitly specify order. Among users who care about how the order of their system tray icons is alphabetical order, rather than by type or usage, such an overwhelmingly popular default that it justifies removing the flexibility that would allow users to make any other choice?

**Comment by Airblader** [comment]
> Since all the applets for persistent services would be launched well in advance from the i3wm config file, at least in my use case, I don't think this would present a problem.

I think you misunderstand. It very much would be. If the processes are started all together, or prior to i3bar, then the order in which they will be picked up will be random.

You'd have to start i3(bar). Then an applet. Wait for it to appear. Start the next one. Wait for it to appear. And so on.

**Comment by Airblader** [comment]
> rather than by type or usage

These are unfortunately all metrics we cannot track to my knowledge. And even with type you'd still have the problem that it'd be optionated which type goes where.

**Comment by infoseek12** [comment]
>I think you misunderstand. It very much would be. If the processes are started all together, or prior to i3bar, then the order in which they will be picked up will be random.

Ahh, I didn't realize that. I think it would, at least, address my proximate issue and get all the persistent services to the right of the transient ones. But I can see how the random ordering of icons would be a problem for a lot of people. I guess it's a minor annoyance of the type that's easier to live with than to fix. Thank you for explaining your design decisions and the technical constraints.

**Comment by Airblader** [comment]
Yeah, as a window manager we don't care about the underlying process that created the window or when it was started — just about the X windows and when they get mapped. But different processes will take different amounts of time to create their window.

I'm going to close this now since the only option is a full configuration for the user defining an order of test icons by hand, which is something we simply don't really want to add. Thanks for your suggestions though!

**Comment by markstos** [comment]
I want to add an case to clarify a use for more control over the bar show/hide behavior:

Sometimes it seems several keys have stopped working and I can't initially figure out why. The problem is that I've activated a "mode" by accident, but the bar is hidden so there's no indication that the mode has changed.

So it's desirable to have an option for the bar in "hide" mode to appear as long as a non-default mode is activate. If the bar  "mode" change could be toggled via IPC, that would make it possible.

There some workarounds for this case:

  * I could bind all the remaining keys to quit the mode I get stuck on. (ex: "Press any other key to quit the mode").
  * I could update the keybindings to enter/exit the mode to also launch/quit a "nag" window as an alternate form of persistent warning.

While the workarounds are options, the most convenient would be a mode that works like "hide" but is really "hide-during-default-mode".

**Comment by markstos** [comment]
I'm following up to say I'm satisfied with the workaround of activating/killing a Nagbar while my non-default mode is active. Here's the syntax I used (for Sway) to define some variables for entering and leaving the mode:

```
set $enterResize exec swaymsg 'mode "Resize / Move Mode"' && swaynag -m 'Resize / Move Mode Active'
set $leaveResize exec swaymsg mode "default" && killall swaynag
```

It's impossible to miss the nag warning!
